src/cargo/sources/registry/mod.rs

```diff
@@ -564,6 +630,15 @@ impl<'cfg> RegistrySource<'cfg> {
 }
 
 impl<'cfg> Source for RegistrySource<'cfg> {
+    fn prefetch(
+        &mut self,
+        deps: &mut dyn ExactSizeIterator<Item = Cow<'_, Dependency>>,
+    ) -> CargoResult<()> {
+        // TODO: conditional index update?
```

Comment:
This one I'm really not sure how to deal with. The logic in `query` is fairly involved here, and I'm not sure how to nicely translate it into an operation over a _sequence_ of deps. Ideas welcome, otherwise I'll just keep tinkering.

Reply:
This is not needed for basic benchmarking, but is a very important optimization for some use cases. Some people really care that `cargo build` does *no* network requests if the lockfile is still valid. This is an ugly backchannel hack to help make that happen. Before stabilization `rfc+` registries also need this property. Ether by implementing this optimization or by finding a better way and removing the hack from all the registry implementations.

Reply:
Ah, so, to be clear, what this comment is alluding to is the fact that `query` _does_ do an index update if a dependency is resolved, whereas `prefetch` _does not_. And the question is whether we _should_ do an index update if prefetching fails. It's always safe not to do so, since we may just end up doing an index update later, but it may be more efficient to have the prefetching also do the index update if needed.